man/hub.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "HUB" "1" "November 2016" "GITHUB" "Hub Manual"
.
.SH "NAME"
\fBhub\fR \- git + hub = github
.
.SH "SYNOPSIS"
\fBhub\fR [\fB\-\-noop\fR] \fICOMMAND\fR \fIOPTIONS\fR
.
.br
\fBhub alias\fR [\fB\-s\fR] [\fISHELL\fR]
.
.SS "Expanded git commands:"
\fBgit init \-g\fR \fIOPTIONS\fR
.
.br
\fBgit clone\fR [\fB\-p\fR] \fIOPTIONS\fR [\fIUSER\fR/]\fIREPOSITORY\fR \fIDIRECTORY\fR
.
.br
\fBgit remote add\fR [\fB\-p\fR] \fIOPTIONS\fR \fIUSER\fR[/\fIREPOSITORY\fR]
.
.br
\fBgit remote set\-url\fR [\fB\-p\fR] \fIOPTIONS\fR \fIREMOTE\-NAME\fR \fIUSER\fR[/\fIREPOSITORY\fR]
.
.br
\fBgit fetch\fR \fIUSER\-1\fR,[\fIUSER\-2\fR,\.\.\.]
.
.br
\fBgit checkout\fR \fIPULLREQ\-URL\fR [\fIBRANCH\fR]
.
.br
\fBgit merge\fR \fIPULLREQ\-URL\fR
.
.br
\fBgit cherry\-pick\fR \fIGITHUB\-REF\fR
.
.br
\fBgit am\fR \fIGITHUB\-URL\fR
.
.br
\fBgit apply\fR \fIGITHUB\-URL\fR
.
.br
\fBgit push\fR \fIREMOTE\-1\fR,\fIREMOTE\-2\fR,\.\.\.,\fIREMOTE\-N\fR [\fIREF\fR]
.
.br
\fBgit submodule add\fR [\fB\-p\fR] \fIOPTIONS\fR [\fIUSER\fR/]\fIREPOSITORY\fR \fIDIRECTORY\fR
.
.SS "Custom git commands:"
\fBgit create\fR [\fINAME\fR] [\fB\-p\fR] [\fB\-d\fR \fIDESCRIPTION\fR] [\fB\-h\fR \fIHOMEPAGE\fR]
.
.br
\fBgit browse\fR [\fB\-u\fR] [[\fIUSER\fR\fB/\fR]\fIREPOSITORY\fR] [SUBPAGE]
.
.br
\fBgit compare\fR [\fB\-u\fR] [\fIUSER\fR] [[\fISTART\fR\.\.\.]\fIEND\fR]
.
.br
\fBgit fork\fR [\fB\-\-no\-remote\fR]
.
.br
\fBgit pull\-request\fR [\fB\-o\fR|\fB\-\-browse\fR] [\fB\-f\fR] [\fB\-m\fR \fIMESSAGE\fR|\fB\-F\fR \fIFILE\fR|\fB\-i\fR \fIISSUE\fR|\fIISSUE\-URL\fR] [\fB\-b\fR \fIBASE\fR] [\fB\-h\fR \fIHEAD\fR]
.
.br
\fBgit ci\-status\fR [\fB\-v\fR] [\fICOMMIT\fR]
.
.SH "DESCRIPTION"
hub enhances various git commands to ease most common workflows with GitHub\.
.
.TP
\fBhub \-\-noop\fR \fICOMMAND\fR
Shows which command(s) would be run as a result of the current command\. Doesn\'t perform anything\.
.
.TP
\fBhub alias\fR [\fB\-s\fR] [\fISHELL\fR]
Shows shell instructions for wrapping git\. If given, \fISHELL\fR specifies the type of shell; otherwise defaults to the value of SHELL environment variable\. With \fB\-s\fR, outputs shell script suitable for \fBeval\fR\.
.
.TP
\fBgit init\fR \fB\-g\fR \fIOPTIONS\fR
Create a git repository as with git\-init(1) and add remote \fBorigin\fR at "git@github\.com:\fIUSER\fR/\fIREPOSITORY\fR\.git"; \fIUSER\fR is your GitHub username and \fIREPOSITORY\fR is the current working directory\'s basename\.
.
.TP
\fBgit clone\fR [\fB\-p\fR] \fIOPTIONS\fR [\fIUSER\fR\fB/\fR]\fIREPOSITORY\fR \fIDIRECTORY\fR
Clone repository "git://github\.com/\fIUSER\fR/\fIREPOSITORY\fR\.git" into \fIDIRECTORY\fR as with git\-clone(1)\. When \fIUSER\fR/ is omitted, assumes your GitHub login\.
.
.IP
If the repository is private or the current user has push access to the repository, hub will use the ssh protocol for cloning\. Use \fB\-p\fR to select the ssh protocol unconditionally\. HTTPS protocol can be used instead by setting "hub\.protocol" (see \fICONFIGURATION\fR)\.
.
.TP
\fBgit remote add\fR [\fB\-p\fR] \fIOPTIONS\fR \fIUSER\fR[\fB/\fR\fIREPOSITORY\fR]
Add remote "git://github\.com/\fIUSER\fR/\fIREPOSITORY\fR\.git" as with git\-remote(1)\. When /\fIREPOSITORY\fR is omitted, the basename of the current working directory is used\. With \fB\-p\fR, use private remote "git@github\.com:\fIUSER\fR/\fIREPOSITORY\fR\.git"\. If \fIUSER\fR is "origin" then uses your GitHub login\.
.
.TP
\fBgit remote set\-url\fR [\fB\-p\fR] \fIOPTIONS\fR \fIREMOTE\-NAME\fR \fIUSER\fR[/\fIREPOSITORY\fR]
Sets the url of remote \fIREMOTE\-NAME\fR using the same rules as \fBgit remote add\fR\.
.
.TP
\fBgit fetch\fR \fIUSER\-1\fR,[\fIUSER\-2\fR,\.\.\.]
Adds missing remote(s) with \fBgit remote add\fR prior to fetching\. New remotes are only added if they correspond to valid forks on GitHub\.
.
.TP
\fBgit checkout\fR \fIPULLREQ\-URL\fR [\fIBRANCH\fR]
Checks out the head of the pull request as a local branch, to allow for reviewing, rebasing and otherwise cleaning up the commits in the pull request before merging\. The name of the local branch can explicitly be set with \fIBRANCH\fR\.
.
.TP
\fBgit merge\fR \fIPULLREQ\-URL\fR
Merge the pull request with a commit message that includes the pull request ID and title, similar to the GitHub Merge Button\.
.
.TP
\fBgit cherry\-pick\fR \fIGITHUB\-REF\fR
Cherry\-pick a commit from a fork using either full URL to the commit or GitHub\-flavored Markdown notation, which is \fBuser@sha\fR\. If the remote doesn\'t yet exist, it will be added\. A \fBgit fetch <user>\fR is issued prior to the cherry\-pick attempt\.
.
.TP
\fBgit [am|apply]\fR \fIGITHUB\-URL\fR
Downloads the patch file for the pull request or commit at the URL and applies that patch from disk with \fBgit am\fR or \fBgit apply\fR\. Similar to \fBcherry\-pick\fR, but doesn\'t add new remotes\. \fBgit am\fR creates commits while preserving authorship info while \fBapply\fR only applies the patch to the working copy\.
.
.TP
\fBgit push\fR \fIREMOTE\-1\fR,\fIREMOTE\-2\fR,\.\.\.,\fIREMOTE\-N\fR [\fIREF\fR]
Push \fIREF\fR to each of \fIREMOTE\-1\fR through \fIREMOTE\-N\fR by executing multiple \fBgit push\fR commands\.
.
.TP
\fBgit submodule add\fR [\fB\-p\fR] \fIOPTIONS\fR [\fIUSER\fR/]\fIREPOSITORY\fR \fIDIRECTORY\fR
Submodule repository "git://github\.com/\fIUSER\fR/\fIREPOSITORY\fR\.git" into \fIDIRECTORY\fR as with git\-submodule(1)\. When \fIUSER\fR/ is omitted, assumes your GitHub login\. With \fB\-p\fR, use private remote "git@github\.com:\fIUSER\fR/\fIREPOSITORY\fR\.git"\.
.
.TP
\fBgit help\fR
Display enhanced git\-help(1)\.
.
.P
hub also adds some custom commands that are otherwise not present in git:
.
.TP
\fBgit create\fR [\fINAME\fR] [\fB\-p\fR] [\fB\-d\fR \fIDESCRIPTION\fR] [\fB\-h\fR \fIHOMEPAGE\fR]
Create a new public GitHub repository from the current git repository and add remote \fBorigin\fR at "git@github\.com:\fIUSER\fR/\fIREPOSITORY\fR\.git"; \fIUSER\fR is your GitHub username and \fIREPOSITORY\fR is the current working directory name\. To explicitly name the new repository, pass in \fINAME\fR, optionally in \fIORGANIZATION\fR/\fINAME\fR form to create under an organization you\'re a member of\. With \fB\-p\fR, create a private repository, and with \fB\-d\fR and \fB\-h\fR set the repository\'s description and homepage URL, respectively\.
.
.TP
\fBgit browse\fR [\fB\-u\fR] [[\fIUSER\fR\fB/\fR]\fIREPOSITORY\fR] [SUBPAGE]
Open repository\'s GitHub page in the system\'s default web browser using \fBopen(1)\fR or the \fBBROWSER\fR env variable\. If the repository isn\'t specified, \fBbrowse\fR opens the page of the repository found in the current directory\. If SUBPAGE is specified, the browser will open on the specified subpage: one of "wiki", "commits", "issues" or other (the default is "tree")\. A specific commit in the default repository can be opened with \fBgit browse \-\- commit/<COMMIT>\fR\. With \fB\-u\fR, outputs the URL rather than opening the browser\.
.
.TP
\fBgit compare\fR [\fB\-u\fR] [\fIUSER\fR] [[\fISTART\fR\.\.\.]\fIEND\fR]
Open a GitHub compare view page in the system\'s default web browser\. \fISTART\fR to \fIEND\fR are branch names, tag names, or commit SHA1s specifying the range of history to compare\. If a range with two dots (\fBa\.\.b\fR) is given, it will be transformed into one with three dots\. If \fISTART\fR is omitted, GitHub will compare against the base branch (the default is "master")\. If \fIEND\fR is omitted, GitHub compare view is opened for the current branch\. With \fB\-u\fR, outputs the URL rather than opening the browser\.
.
.TP
\fBgit fork\fR [\fB\-\-no\-remote\fR]
Forks the original project (referenced by "origin" remote) on GitHub and adds a new remote for it under your username\.
.
.TP
\fBgit pull\-request\fR [\fB\-o\fR|\fB\-\-browse\fR] [\fB\-f\fR] [\fB\-m\fR \fIMESSAGE\fR|\fB\-F\fR \fIFILE\fR|\fB\-i\fR \fIISSUE\fR|\fIISSUE\-URL\fR] [\fB\-b\fR \fIBASE\fR] [\fB\-h\fR \fIHEAD\fR]
Opens a pull request on GitHub for the project that the "origin" remote points to\. The default head of the pull request is the current branch\. Both base and head of the pull request can be explicitly given in one of the following formats: "branch", "owner:branch", "owner/repo:branch"\. This command will abort operation if it detects that the current topic branch has local commits that are not yet pushed to its upstream branch on the remote\. To skip this check, use \fB\-f\fR\.
.
.IP
Without \fIMESSAGE\fR or \fIFILE\fR, a text editor will open in which title and body of the pull request can be entered in the same manner as git commit message\. Pull request message can also be passed via stdin with \fB\-F \-\fR\.
.
.IP
With \fB\-o\fR or \fB\-\-browse\fR, the new pull request will open in the web browser\.
.
.IP
Issue to pull request conversion via \fB\-i <ISSUE>\fR or \fIISSUE\-URL\fR arguments is deprecated and will likely be removed from the future versions of both hub and GitHub API\.
.
.TP
\fBgit ci\-status\fR [\fB\-v\fR] [\fICOMMIT\fR]
Looks up the SHA for \fICOMMIT\fR in GitHub Status API and displays the latest status\. Exits with one of:
.
.br
success (0), error (1), failure (1), pending (2), no status (3)
.
.IP
If \fB\-v\fR is given, additionally print detailed report of all checks and their URLs\.
.
.SH "CONFIGURATION"
.
.SS "GitHub OAuth authentication"
Hub will prompt for GitHub username & password the first time it needs to access the API and exchange it for an OAuth token, which it saves in "~/\.config/hub"\.
.
.P
To avoid being prompted, use \fIGITHUB_USER\fR and \fIGITHUB_PASSWORD\fR environment variables\.
.
.P
Alternatively, you may provide \fIGITHUB_TOKEN\fR, an access token with \fIrepo\fR permissions\. This will not be written to \fB~/\.config/hub\fR\.
.
.SS "HTTPS instead of git protocol"
If you prefer the HTTPS protocol for GitHub repositories, you can set "hub\.protocol" to "https"\. This will affect \fBclone\fR, \fBfork\fR, \fBremote add\fR and other operations that expand references to GitHub repositories as full URLs that otherwise use git and ssh protocols\.
.
.P
"hub\.protocol" only applies when the "OWNER/REPO" shorthand is used instead of a full git URL\.
.
.IP "" 4
.
.nf

# default behavior
$ git clone defunkt/repl
< git clone >

# opt into HTTPS:
$ git config \-\-global hub\.protocol https
$ git clone defunkt/repl
< https clone >
.
.fi
.
.IP "" 0
.
.SS "GitHub Enterprise"
By default, hub will only work with repositories that have remotes which point to github\.com\. GitHub Enterprise hosts need to be whitelisted to configure hub to treat such remotes same as github\.com:
.
.IP "" 4
.
.nf

$ git config \-\-global \-\-add hub\.host my\.git\.org
.
.fi
.
.IP "" 0
.
.P
The default host for commands like \fBinit\fR and \fBclone\fR is still github\.com, but this can be affected with the \fBGITHUB_HOST\fR environment variable:
.
.IP "" 4
.
.nf

$ GITHUB_HOST=my\.git\.org git clone myproject
.
.fi
.
.IP "" 0
.
.SH "EXAMPLES"
.
.SS "git clone"
.
.nf

$ git clone schacon/ticgit
> git clone git://github\.com/schacon/ticgit\.git

$ git clone \-p schacon/ticgit
> git clone git@github\.com:schacon/ticgit\.git

$ git clone resque
> git clone git@github\.com:YOUR_USER/resque\.git
.
.fi
.
.SS "git remote add"
.
.nf

$ git remote add rtomayko
> git remote add rtomayko git://github\.com/rtomayko/CURRENT_REPO\.git

$ git remote add \-p rtomayko
> git remote add rtomayko git@github\.com:rtomayko/CURRENT_REPO\.git

$ git remote add origin
> git remote add origin git://github\.com/YOUR_USER/CURRENT_REPO\.git
.
.fi
.
.SS "git fetch"
.
.nf

$ git fetch mislav
> git remote add mislav git://github\.com/mislav/REPO\.git
> git fetch mislav

$ git fetch mislav,xoebus
> git remote add mislav \.\.\.
> git remote add xoebus \.\.\.
> git fetch \-\-multiple mislav xoebus
.
.fi
.
.SS "git cherry\-pick"
.
.nf

$ git cherry\-pick http://github\.com/mislav/REPO/commit/SHA
> git remote add \-f \-\-no\-tags mislav git://github\.com/mislav/REPO\.git
> git cherry\-pick SHA

$ git cherry\-pick mislav@SHA
> git remote add \-f \-\-no\-tags mislav git://github\.com/mislav/CURRENT_REPO\.git
> git cherry\-pick SHA

$ git cherry\-pick mislav@SHA
> git fetch mislav
> git cherry\-pick SHA
.
.fi
.
.SS "git am, git apply"
.
.nf

$ git am https://github\.com/defunkt/hub/pull/55
[ downloads patch via API ]
> git am /tmp/55\.patch

$ git am \-\-ignore\-whitespace https://github\.com/davidbalbert/hub/commit/fdb9921
[ downloads patch via API ]
> git am \-\-ignore\-whitespace /tmp/fdb9921\.patch

$ git apply https://gist\.github\.com/8da7fb575debd88c54cf
[ downloads patch via API ]
> git apply /tmp/gist\-8da7fb575debd88c54cf\.txt
.
.fi
.
.SS "git fork"
.
.nf

$ git fork
[ repo forked on GitHub ]
> git remote add \-f YOUR_USER git@github\.com:YOUR_USER/CURRENT_REPO\.git
.
.fi
.
.SS "git pull\-request"
.
.nf

# while on a topic branch called "feature":
$ git pull\-request
[ opens text editor to edit title & body for the request ]
[ opened pull request on GitHub for "YOUR_USER:feature" ]

# explicit title, pull base & head:
$ git pull\-request \-m "Implemented feature X" \-b defunkt:master \-h mislav:feature
.
.fi
.
.SS "git checkout"
.
.nf

$ git checkout https://github\.com/defunkt/hub/pull/73
> git remote add \-f \-\-no\-tags \-t feature mislav git://github\.com/mislav/hub\.git
> git checkout \-\-track \-B mislav\-feature mislav/feature

$ git checkout https://github\.com/defunkt/hub/pull/73 custom\-branch\-name
.
.fi
.
.SS "git merge"
.
.nf

$ git merge https://github\.com/defunkt/hub/pull/73
> git fetch git://github\.com/mislav/hub\.git +refs/heads/feature:refs/remotes/mislav/feature
> git merge mislav/feature \-\-no\-ff \-m \'Merge pull request #73 from mislav/feature\.\.\.\'
.
.fi
.
.SS "git create"
.
.nf

$ git create
[ repo created on GitHub ]
> git remote add origin git@github\.com:YOUR_USER/CURRENT_REPO\.git

# with description:
$ git create \-d \'It shall be mine, all mine!\'

$ git create recipes
[ repo created on GitHub ]
> git remote add origin git@github\.com:YOUR_USER/recipes\.git

$ git create sinatra/recipes
[ repo created in GitHub organization ]
> git remote add origin git@github\.com:sinatra/recipes\.git
.
.fi
.
.SS "git init"
.
.nf

$ git init \-g
> git init
> git remote add origin git@github\.com:YOUR_USER/REPO\.git
.
.fi
.
.SS "git push"
.
.nf

$ git push origin,staging,qa bert_timeout
> git push origin bert_timeout
> git push staging bert_timeout
> git push qa bert_timeout
.
.fi
.
.SS "git browse"
.
.nf

$ git browse
> open https://github\.com/YOUR_USER/CURRENT_REPO

$ git browse \-\- commit/SHA
> open https://github\.com/YOUR_USER/CURRENT_REPO/commit/SHA

$ git browse \-\- issues
> open https://github\.com/YOUR_USER/CURRENT_REPO/issues

$ git browse \-\- issues/10
> open https://github\.com/YOUR_USER/CURRENT_REPO/issues/10

$ git browse schacon/ticgit
> open https://github\.com/schacon/ticgit


HISTORY.md
## master

* change `pull-request` interface to allow passing title & body
* avoid hard breaks in `pull-request` message authored in Vim
* save and reuse `pull-request` message if creating it failed
* new `ci-status` command for checking GitHub Status API

## 1.10.6 (2013-04-25)

* compensate for new GitHub API enforcement
* fix asking for password on Ruby <= 1.8.6
* fix `fetch` for forks where username contains a dash

## 1.10.5 (2013-03-02)

* helpful `pull-request` error message when base project is invalid
* fix `compare` for ranges containing "owner:branch" notation
* enable `--name` argument for `submodule add`

## 1.10.4 (2012-12-29)

* fixes for Windows
* display more validation errors on GitHub API failures
* persist correctly capitalized GitHub login name

## 1.10.3 (2012-11-22)

* fix `browse` on Windows
* fix hub on JRuby
* try fixing Ruby 1.8.7 error with API requests
* fix various edge cases around `create` command
* have `fork` set up a remote even if own fork already exists
* fix `pull-request` with Unicode chars on Ruby 1.8
* respect $GITHUB_USER & $GITHUB_PASSWORD
* improve reading git remotes configuration

## 1.10.2 (2012-07-24)

* fix pushing multiple refs to multiple remotes
* support ssh.github.com
* security: mode for ~/.config/hub is 0600
* fix integration with GitHub Enterprise API
* fix cloning repos that start with a period

## 1.10.1 (2012-05-28)

* don't choke on empty config file
* fix `browse` when not in git repo

## 1.10.0 (2012-05-08)

* improve improved help text
* fix GitHub username prompt in `create` command
* make `fetch` command work with private repos too
* add `merge` command to merge pull requests

## 1.9.0 (2012-05-04)

* internal refactoring and code reorganization
* switch to GitHub API v3 and authenticate via OAuth
* auth info is now stored in ~/.config/hub instead of ~/.gitconfig

## 1.8.4 (2012-03-20)

* add bash, zsh completion
* improve `hub alias` command
* change `git fork` so it fails when repo already exists under user
* teach custom commands to respect `-h` & `--help` flags
* `pull-request`: better error message for invalid remotes/URLs
* respect local SSH aliases for host names

## 1.8.3 (2012-03-02)

* fix `pull-request` from branch tracking another local branch
* fix `browse` command when not on any branch

## 1.8.2 (2012-02-07)

* if `pull-request` editor is vim, set appropriate filetype
* `pull-request` editor message defaults to single commit message
* fix cherry-picking from an existing remote
* fix `clone` from local repository
* `checkout` command forwards flags to internal checkout command,
  force-resets the existing local branch by default
* fix `am` command when given URLs that include the fragment

## 1.8.1 (2012-01-24)

* fix JSON parsing error while using GitHub API
* HTTP(S) proxy support

## 1.8.0 (2012-01-03)

* fix `pull-request` on GH Enterprise project branch without upstream
* ensure Content-Length for POST requests
* handle pull requests from private repos
* support branches with slashes in their name
* display server errors when creating pullrequest fails
* support GitHub Enterprise via multiple whitelisted host names
* GitHub remote urls don't have to necessarily end in ".git"
* fix `git init -g`
* authenticate all API requests, helps dealing with private repos
* ensure periods are allowed in repository names
* fix am/apply commands if TMPDIR environment variable isn't set

## 1.7.0 (2011-11-24)

* lock down standalone script to system ruby
* don't try to use command output pager on Windows
* opt in for HTTPS: `git config hub.protocol https`
* add `hub pull-request`
* improve detecting upstream configuration ("tracking" branches)
* introduce `hub --noop`
* `hub apply` now downloads GitHub patches same as `hub am`
* `hub create <name>` to explicitly name a repository
* switch API communication to HTTPS
* better handling of API HTTP exceptions
* replace two dots (`sha1..sha2`) with three for ranges in `compare`
* avoid ugly error & stack trace when git is not found on the system

## 1.6.1 (2011-05-13)

* `git push remote1,remote2` without branch name pushes the current branch
* fix `browse` command for current repo with no tracking setup
* preserve global flags to git such as `--bare` and `--git-dir=/some/path`
* true cross-platform command detection and browser launcher

## 1.6.0 (2011-03-24)

* `am` strips extra path from pull reqs URLs such as "pull/42/files"
* Fixed permissions on `hub(1)` when installing
* gem renamed from `git-hub` to `hub`!

## 1.5.1 (2011-03-18)

* support git aliases
* Bugfix: `browse/compare` for wiki repos
* gracefully handle HTTP errors in `create` and `fork`
* `hub am` supports Gist URLs
* Bugfix: `clone` command doesn't get confused by mixed arguments

## 1.5.0 (2010-12-28)

* compensate for GitHub switch to HTTPS
* `hub am`: cherry-pick pull request and commit URLs
* support multiple URLs for a single remote
* Bugfix: ensure that internal ruby methods can't pretend to be git commands
* Bugfix: don't show help when `--exec-path` or `--html-path` flags are used
* Support for `GITHUB_USER` and `GITHUB_TOKEN` env variables
* Eliminate some ruby warnings

## 1.4.1 (2010-08-08)
## 1.4.0 (2010-08-08)

* Added new `hub create` command
* Added support for `remote set-url`
* Bugfix: Don't try multiple git commands on a non-git dir when grabbing remote
* Bugfix: Adding remotes when no remotes exist

## 1.3.2 (2010-07-24)

* bugfix: cherry picking of commit URL
* bugfix: git init -g

## 1.3.1 (2010-04-29)
## 1.3.0 (2010-04-29)

* Tracking branches awareness
* `git browse` subpages (e.g. `git browse repo issues`)
* `git fetch <fork>` adds new remotes if missing
* `cherry-pick` supports GitHub commit URLs and "user@sha" notation

## 1.2.0 (2010-04-11)

* `hub compare` command - Thanks joshthecoder!

## 1.1.0 (2010-04-07)

* `hub fork` command - Thanks Mislav!

## 1.0.3 (2010-03-10)

* Bugfix: `hub remote` for repos with -, /, etc

## 1.0.2 (2010-03-07)


etc/hub.bash_completion.sh
# hub tab-completion script for bash.
# This script complements the completion script that ships with git.

# If there is no git tab completion, but we have the _completion loader try to load it
if ! declare -F _git > /dev/null && declare -F _completion_loader > /dev/null; then
  _completion_loader git
fi

# Check that git tab completion is available
if declare -F _git > /dev/null; then
  # Duplicate and rename the 'list_all_commands' function
  eval "$(declare -f __git_list_all_commands | \
        sed 's/__git_list_all_commands/__git_list_all_commands_without_hub/')"

  # Wrap the 'list_all_commands' function with extra hub commands
  __git_list_all_commands() {
    cat <<-EOF
alias
pull-request
fork
create
browse
compare
ci-status
EOF
    __git_list_all_commands_without_hub
  }

  # Ensure cached commands are cleared
  __git_all_commands=""

  ##########################
  # hub command completions
  ##########################

  # hub alias [-s] [SHELL]
  _git_alias() {
    local i c=2 s=-s sh shells="bash zsh sh ksh csh fish"
    while [ $c -lt $cword ]; do
      i="${words[c]}"
      case "$i" in
        -s)
          unset s
          ;;
        *)
          for sh in $shells; do
            if [ "$sh" = "$i" ]; then
              unset shells
              break
            fi
          done
          ;;
      esac
      ((c++))
    done
    __gitcomp "$s $shells"
  }

  # hub browse [-u] [--|[USER/]REPOSITORY] [SUBPAGE]
  _git_browse() {
    local i c=2 u=-u repo subpage
    local subpages_="commits issues tree wiki pulls branches stargazers
      contributors network network/ graphs graphs/"
    local subpages_network="members"
    local subpages_graphs="commit-activity code-frequency punch-card"
    while [ $c -lt $cword ]; do
      i="${words[c]}"
      case "$i" in
        -u)
          unset u
          ;;
        *)
          if [ -z "$repo" ]; then
            repo=$i
          else
            subpage=$i
          fi
          ;;
      esac
      ((c++))
    done
    if [ -z "$repo" ]; then
      __gitcomp "$u -- $(__hub_github_repos '\p')"
    elif [ -z "$subpage" ]; then
      case "$cur" in
        */*)
          local pfx="${cur%/*}" cur_="${cur#*/}"
          local subpages_var="subpages_$pfx"
          __gitcomp "${!subpages_var}" "$pfx/" "$cur_"
          ;;
        *)
          __gitcomp "$u ${subpages_}"
          ;;
      esac
    else
      __gitcomp "$u"
    fi
  }

  # hub compare [-u] [USER[/REPOSITORY]] [[START...]END]
  _git_compare() {
    local i c=$((cword - 1)) u=-u user remote owner repo arg_repo rev
    while [ $c -gt 1 ]; do
      i="${words[c]}"
      case "$i" in
        -u)
          unset u
          ;;
        *)
          if [ -z "$rev" ]; then
            # Even though the logic below is able to complete both user/repo
            # and revision in the right place, when there is only one argument
            # (other than -u) in the command, that argument will be taken as
            # revision. For example:
            # $ hub compare -u upstream
            # > https://github.com/USER/REPO/compare/upstream
            if __hub_github_repos '\p' | grep -Eqx "^$i(/[^/]+)?"; then
              arg_repo=$i
            else
              rev=$i
            fi
          elif [ -z "$arg_repo" ]; then
            arg_repo=$i
          fi
          ;;
      esac
      ((c--))
    done

    # Here we want to find out the git remote name of user/repo, in order to
    # generate an appropriate revision list
    if [ -z "$arg_repo" ]; then
      user=$(__hub_github_user)
      if [ -z "$user" ]; then
        for i in $(__hub_github_repos); do
          remote=${i%%:*}
          repo=${i#*:}
          if [ "$remote" = origin ]; then
            break
          fi
        done
      else
        for i in $(__hub_github_repos); do
          remote=${i%%:*}
          repo=${i#*:}
          owner=${repo%%/*}
          if [ "$user" = "$owner" ]; then
            break
          fi
        done
      fi
    else
      for i in $(__hub_github_repos); do
        remote=${i%%:*}
        repo=${i#*:}
        owner=${repo%%/*}
        case "$arg_repo" in
          "$repo"|"$owner")
            break
            ;;
        esac
      done
    fi

    local pfx cur_="$cur"
    case "$cur_" in
      *..*)
        pfx="${cur_%%..*}..."
        cur_="${cur_##*..}"
        __gitcomp_nl "$(__hub_revlist $remote)" "$pfx" "$cur_"
        ;;
      *)
        if [ -z "${arg_repo}${rev}" ]; then
          __gitcomp "$u $(__hub_github_repos '\o\n\p') $(__hub_revlist $remote)"
        elif [ -z "$rev" ]; then
          __gitcomp "$u $(__hub_revlist $remote)"
        else
          __gitcomp "$u"
        fi
        ;;
    esac
  }

  # hub create [NAME] [-p] [-d DESCRIPTION] [-h HOMEPAGE]
  _git_create() {
    local i c=2 name repo flags="-p -d -h"
    while [ $c -lt $cword ]; do
      i="${words[c]}"
      case "$i" in
        -d|-h)
          ((c++))
          flags=${flags/$i/}
          ;;
        -p)
          flags=${flags/$i/}
          ;;
        *)
          name=$i
          ;;
      esac
      ((c++))
    done
    if [ -z "$name" ]; then
      repo=$(basename "$(pwd)")
    fi
    case "$prev" in
      -d|-h)
        COMPREPLY=()
        ;;
      -p|*)
        __gitcomp "$repo $flags"
        ;;
    esac
  }

  # hub fork [--no-remote]
  _git_fork() {
    local i c=2 remote=yes
    while [ $c -lt $cword ]; do
      i="${words[c]}"
      case "$i" in
        --no-remote)
          unset remote
          ;;
      esac
      ((c++))
    done
    if [ -n "$remote" ]; then
      __gitcomp "--no-remote"
    fi
  }

  # hub pull-request [-f] [-m <MESSAGE>|-F <FILE>|-i <ISSUE>|<ISSUE-URL>] [-b <BASE>] [-h <HEAD>] [-a <USER>] [-M <MILESTONE>] [-l <LABELS>]
  _git_pull_request() {
    local i c=2 flags="-f -m -F -i -b -h -a -M -l"
    while [ $c -lt $cword ]; do
      i="${words[c]}"
      case "$i" in
        -m|-F|-i|-b|-h|-a|-M|-l)
          ((c++))
          flags=${flags/$i/}
          ;;
        -f)
          flags=${flags/$i/}
          ;;
      esac
      ((c++))
    done
    case "$prev" in
      -i)
        COMPREPLY=()
        ;;
      -b|-h|-a|-M|-l)
        # (Doesn't seem to need this...)
        # Uncomment the following line when 'owner/repo:[TAB]' misbehaved
        #_get_comp_words_by_ref -n : cur
        __gitcomp_nl "$(__hub_heads)"
        # __ltrim_colon_completions "$cur"
        ;;
      -F)
        COMPREPLY=( "$cur"* )
        ;;
      -f|*)
        __gitcomp "$flags"
        ;;
    esac
  }

  ###################
  # Helper functions
  ###################

  # __hub_github_user [HOST]
  # Return $GITHUB_USER or the default github user defined in hub config
  # HOST - Host to be looked-up in hub config. Default is "github.com"
  __hub_github_user() {
    if [ -n "$GITHUB_USER" ]; then
      echo $GITHUB_USER
      return
    fi
    local line h k v host=${1:-github.com} config=${HUB_CONFIG:-~/.config/hub}
    if [ -f "$config" ]; then
      while read line; do
        if [ "$line" = "---" ]; then
          continue
        fi
        k=${line%%:*}
        v=${line#*:}
        if [ -z "$v" ]; then
          if [ "$h" = "$host" ]; then
            break
          fi
          h=$k
          continue
        fi
        k=${k#* }
        v=${v#* }
        if [ "$h" = "$host" ] && [ "$k" = "user" ]; then
          echo "$v"
          break
        fi
      done < "$config"
    fi
  }

  # __hub_github_repos [FORMAT]
  # List all github hosted repository
  # FORMAT - Format string contains multiple of these:
  #   \m  remote
  #   \p  owner/repo
  #   \o  owner
  #   escaped characters (\n, \t ...etc) work
  # If omitted, prints all github repos in the format of "remote:owner/repo"
  __hub_github_repos() {
    local f format=$1
    if [ -z "$(__gitdir)" ]; then
      return
    fi
    if [ -z "$format" ]; then
      format='\1:\2'
    else
      format=${format//\m/\1}
      format=${format//\p/\2}
      format=${format//\o/\3}
    fi
    command git config --get-regexp 'remote\.[^.]*\.url' |
    grep -E ' ((https?|git)://|git@)github\.com[:/][^:/]+/[^/]+$' |
    sed -E 's#^remote\.([^.]+)\.url +.+[:/](([^/]+)/[^.]+)(\.git)?$#'"$format"'#'
  }

  # __hub_heads
  # List all local "branch", and remote "owner/repo:branch"
  __hub_heads() {
    local i remote repo branch dir=$(__gitdir)
    if [ -d "$dir" ]; then
      command git --git-dir="$dir" for-each-ref --format='%(refname:short)' \
        "refs/heads/"
      for i in $(__hub_github_repos); do
        remote=${i%%:*}
        repo=${i#*:}
        command git --git-dir="$dir" for-each-ref --format='%(refname:short)' \
          "refs/remotes/${remote}/" | while read branch; do
          echo "${repo}:${branch#${remote}/}"
        done
      done
    fi
  }

  # __hub_revlist [REMOTE]
  # List all tags, and branches under REMOTE, without the "remote/" prefix
  # REMOTE - Remote name to search branches from. Default is "origin"
  __hub_revlist() {
    local i remote=${1:-origin} dir=$(__gitdir)
    if [ -d "$dir" ]; then
      command git --git-dir="$dir" for-each-ref --format='%(refname:short)' \
        "refs/remotes/${remote}/" | while read i; do
        echo "${i#${remote}/}"
      done
      command git --git-dir="$dir" for-each-ref --format='%(refname:short)' \
        "refs/tags/"
    fi
  }

  # Enable completion for hub even when not using the alias
  complete -o bashdefault -o default -o nospace -F _git hub 2>/dev/null \
    || complete -o default -o nospace -F _git hub
fi
